File: week-1/problem-set-recursion/prob4.py
"""Prob4 example.

This file belongs to the week 1 recursion practice solutions. It uses rabbit_count to illustrate the main idea with a small runnable example.
"""

# Fibonacci
def rabbit_count(n):
    """Return the number of rabbit pairs for the classic Fibonacci-style growth model.

    Args:
        n (int): The n input used by this function.

    Returns:
        int: The integer result produced by the algorithm.
    """
    if n == 1 or n == 2:
        return 1
    else:
        return rabbit_count(n-1) + rabbit_count(n-2)

n = 6
print("Rabbit at the year: ", rabbit_count(n)) 

File: week-1/problem-set-recursion/prob4.cpp
/**
 * @file prob4.cpp
 * @brief C++ version of the recursive rabbit-count example.
 */

#include <iostream>

using namespace std;

/**
 * @brief Return the number of rabbit pairs in the Fibonacci-style growth model.
 * @param n int Month or year number in the recurrence relation.
 * @return int Number of rabbit pairs predicted for step n.
 */
int rabbit_count(int n) {
    if (n == 1 || n == 2) {
        return 1;
    }
    return rabbit_count(n - 1) + rabbit_count(n - 2);
}

/**
 * @brief Run the rabbit-count demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    int n = 6;
    cout << "Rabbit at the year: " << rabbit_count(n) << '\n';
    return 0;
}
